File: app/repository/ChallengeRepository.java
package repository;

import models.Challenge;
import play.db.ebean.EbeanConfig;

import javax.inject.Inject;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

/**
 * A repository that executes database operations in a different
 * execution context.
 */
public class ChallengeRepository extends BaseRepository<Challenge> {

    @Inject
    public ChallengeRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext executionContext) {
        super(ebeanConfig, executionContext);
    }

    @Override
    protected Class<Challenge> getModelClass() {
        return Challenge.class;
    }

    /**
     * Return a list of all challenges
     */
    public CompletionStage<List<Challenge>> all() {
        return get(query -> query
                .orderBy("createdAt")
                .select("name,createdAt")
                .fetch("owner", "id,name")
                .findList());
    }

    public CompletionStage<Optional<Challenge>> view(Long id) {
        return get(id, query -> query
                .fetch("entries", "id,currentName")
                .fetch("owner", "id,name")
                .fetch("versions", "id,createdAt,versionId"));
    }

    public CompletionStage<Optional<Challenge>> update(Challenge data) {
        Challenge toSave = new Challenge();
        toSave.id = data.id;
        toSave.name = data.name;
        toSave.refId = data.refId;
        toSave.owner = data.owner;
        return updateModel(data);
    }

}

File: app/repository/ChallengeRepository.scala
package repository

import javax.inject._
import models.Challenge
import play.db.ebean.EbeanConfig

import scala.concurrent.Future

class ChallengeRepository @Inject()(val ebeanConfig: EbeanConfig, val executionContext: DatabaseExecutionContext) extends BaseRepository[Challenge] {
  val modelClass: Class[Challenge] = classOf[Challenge]

  def all(): Future[List[Challenge]] =
    getList {
      query
        .select("name,createdAt")
        .fetch("owner", "id,name")
        //      .where().eq("status", Challenge.Status.Active)
        .orderBy("createdAt")
    }

  def view(id: Long): Future[Option[Challenge]] =
    getOneWhere(id) {
      query
        .fetch("entries", "id,currentName")
        .fetch("owner", "id,name")
        .fetch("versions", "id,createdAt,version")
    }


  def activeChallenges(): Future[List[Challenge]] = getList {
    query
      .select("id,repoUrl,builder,buildParameters,iterationGoal")
      .fetch("versions", "id,gitHash,matchmaker,gameSize,scorer,scoringParameters")
      .fetch("versions.games", "id,startTime,endTime")
      .where().eq("status", Challenge.Status.Active)
      .orderBy("versions.createdAt")
  }

}
